PartitionAssignmentsCorrectnessTest testPartitionAssignments_whenNodesStartedTerminated

- Fails on `Hazelcast-4.maintenance-OracleJDK8`
- Fails on [Build #77 (May 28, 2020 2:16:44 AM)](http://jenkins.hazelcast.com/view/Official%20Builds/job/Hazelcast-4.maintenance-OracleJDK8/77/testReport/junit/com.hazelcast.internal.partition/PartitionAssignmentsCorrectnessTest/testPartitionAssignments_whenNodesStartedTerminated_backups_2_nodes_3_/)

- Error 
```
On Member [127.0.0.1]:5702 - 1b5af6ca-a423-4473-839f-2787cf8fddd4 this, Replica 1 is not found in Partition [9]{
 0:[127.0.0.1]:5702 - 1b5af6ca-a423-4473-839f-2787cf8fddd4
 3:[127.0.0.1]:5701 - 1534ca12-e1df-4e80-bd14-bc56ca07ed6e
}
```
- Stacktrace
```
java.lang.AssertionError: 
On Member [127.0.0.1]:5702 - 1b5af6ca-a423-4473-839f-2787cf8fddd4 this, Replica 1 is not found in Partition [9]{
	0:[127.0.0.1]:5702 - 1b5af6ca-a423-4473-839f-2787cf8fddd4
	3:[127.0.0.1]:5701 - 1534ca12-e1df-4e80-bd14-bc56ca07ed6e
}
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at org.junit.Assert.assertNotNull(Assert.java:712)
	at com.hazelcast.internal.partition.PartitionCorrectnessTestSupport.assertPartitionAssignments(PartitionCorrectnessTestSupport.java:193)
	at com.hazelcast.internal.partition.AbstractPartitionAssignmentsCorrectnessTest$1.run(AbstractPartitionAssignmentsCorrectnessTest.java:83)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1356)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1458)
	at com.hazelcast.internal.partition.AbstractPartitionAssignmentsCorrectnessTest.assertPartitionAssignmentsEventually(AbstractPartitionAssignmentsCorrectnessTest.java:80)
	at com.hazelcast.internal.partition.AbstractPartitionAssignmentsCorrectnessTest.assertPartitionAssignmentsEventually(AbstractPartitionAssignmentsCorrectnessTest.java:76)
	at com.hazelcast.internal.partition.AbstractPartitionAssignmentsCorrectnessTest.testPartitionAssignments_whenNodesStartedTerminated(AbstractPartitionAssignmentsCorrectnessTest.java:48)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:114)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:106)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
```
- Found in logs
```
Caused by: com.hazelcast.spi.exception.RetryableHazelcastException: Cannot set active migration to MigrationInfo{uuid=2ab99685-ff47-4cec-8ec8-27c66be95a27, partitionId=9, source=null, sourceCurrentReplicaIndex=-1, sourceNewReplicaIndex=-1, destination=[127.0.0.1]:5701 - 1534ca12-e1df-4e80-bd14-bc56ca07ed6e, destinationCurrentReplicaIndex=3, destinationNewReplicaIndex=1, master=[127.0.0.1]:5701, initialPartitionVersion=945, partitionVersionIncrement=2, processing=false, status=ACTIVE}. Current active migration is MigrationInfo{uuid=22c84cb8-c0e5-497e-8592-fea4f79dd7fd, partitionId=89, source=[127.0.0.1]:5704 - cbe65351-e53c-47e4-96d9-de2d0929aca6, sourceCurrentReplicaIndex=2, sourceNewReplicaIndex=3, destination=[127.0.0.1]:5702 - 1b5af6ca-a423-4473-839f-2787cf8fddd4, destinationCurrentReplicaIndex=-1, destinationNewReplicaIndex=2, master=[127.0.0.1]:5701, initialPartitionVersion=608, partitionVersionIncrement=2, processing=false, status=INVALID}
	at com.hazelcast.internal.partition.operation.BaseMigrationOperation.setActiveMigration(BaseMigrationOperation.java:225) ~[classes/:?]
	at com.hazelcast.internal.partition.operation.MigrationRequestOperation.call(MigrationRequestOperation.java:88) ~[classes/:?]
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.call(OperationRunnerImpl.java:228) ~[classes/:?]
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:217) ~[classes/:?]
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:433) ~[classes/:?]
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:166) ~[classes/:?]
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:136) ~[classes/:?]
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.run(OperationThread.java:123) ~[classes/:?]
```

kindly check.

There's a serialization issue in the failure but I couldn't figure out yet the cause.

```
08:10:57,634 ERROR |testPartitionData_whenBackupNodesStartedTerminated[backups:3,nodes:4,fragmented:false]| - [OperationExecutorImpl] hz.youthful_tu.generic-operation.thread-11 - [127.0.0.1]:5703 [dev] [4.1-SNAPSHOT] Failed to process: Packet{partitionId=-1, frameLength=10617, conn=MockConnection{localEndpoint=[127.0.0.1]:5703, remoteEndpoint=[127.0.0.1]:5701, alive=true}, rawFlags=10001, isUrgent=true, packetType=OPERATION, typeSpecificFlags=[isResponse=false, isOpControl=false]} on: hz.youthful_tu.generic-operation.thread-11
com.hazelcast.nio.serialization.HazelcastSerializationException: Problem while reading DataSerializable, namespace: -2, ID: 1, class: 'null', exception: Illegal Capacity: -511
	at java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:627) ~[?:1.8.0_251]
	at com.hazelcast.internal.util.ExceptionUtil.tryWrapInSameClass(ExceptionUtil.java:233) ~[classes/:?]
	at com.hazelcast.internal.util.ExceptionUtil.wrapException(ExceptionUtil.java:138) ~[classes/:?]
	at com.hazelcast.internal.util.ExceptionUtil.peel(ExceptionUtil.java:114) ~[classes/:?]
	at com.hazelcast.internal.util.ExceptionUtil.peel(ExceptionUtil.java:73) ~[classes/:?]
	at com.hazelcast.internal.util.ExceptionUtil.rethrow(ExceptionUtil.java:148) ~[classes/:?]
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:459) ~[classes/:?]
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:166) ~[classes/:?]
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:136) [classes/:?]
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.executeRun(OperationThread.java:123) [classes/:?]
	at com.hazelcast.internal.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:102) [classes/:?]
Caused by: com.hazelcast.nio.serialization.HazelcastSerializationException: Problem while reading DataSerializable, namespace: -2, ID: 1, class: 'null', exception: Illegal Capacity: -511
	at com.hazelcast.internal.serialization.impl.DataSerializableSerializer.rethrowReadException(DataSerializableSerializer.java:183) ~[classes/:?]
	at com.hazelcast.internal.serialization.impl.DataSerializableSerializer.readInternal(DataSerializableSerializer.java:164) ~[classes/:?]
	at com.hazelcast.internal.serialization.impl.DataSerializableSerializer.read(DataSerializableSerializer.java:106) ~[classes/:?]
	at com.hazelcast.internal.serialization.impl.DataSerializableSerializer.read(DataSerializableSerializer.java:51) ~[classes/:?]
	at com.hazelcast.internal.serialization.impl.StreamSerializerAdapter.read(StreamSerializerAdapter.java:44) ~[classes/:?]
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.readObject(AbstractSerializationService.java:286) ~[classes/:?]
	at com.hazelcast.internal.serialization.impl.ByteArrayObjectDataInput.readObject(ByteArrayObjectDataInput.java:567) ~[classes/:?]
	at com.hazelcast.internal.cluster.impl.operations.MembersUpdateOp.readInternalImpl(MembersUpdateOp.java:117) ~[classes/:?]
	at com.hazelcast.internal.cluster.impl.operations.FinalizeJoinOp.readInternalImpl(FinalizeJoinOp.java:169) ~[classes/:?]
	at com.hazelcast.internal.cluster.impl.operations.MembersUpdateOp.readInternal(MembersUpdateOp.java:123) ~[classes/:?]
	at com.hazelcast.spi.impl.operationservice.Operation.readData(Operation.java:750) ~[classes/:?]
	at com.hazelcast.internal.serialization.impl.DataSerializableSerializer.readInternal(DataSerializableSerializer.java:160) ~[classes/:?]
	at com.hazelcast.internal.serialization.impl.DataSerializableSerializer.read(DataSerializableSerializer.java:106) ~[classes/:?]
	at com.hazelcast.internal.serialization.impl.DataSerializableSerializer.read(DataSerializableSerializer.java:51) ~[classes/:?]
	at com.hazelcast.internal.serialization.impl.StreamSerializerAdapter.read(StreamSerializerAdapter.java:44) ~[classes/:?]
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toObject(AbstractSerializationService.java:205) ~[classes/:?]
	at com.hazelcast.spi.impl.NodeEngineImpl.toObject(NodeEngineImpl.java:346) ~[classes/:?]
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:437) ~[classes/:?]
	... 4 more
Caused by: java.lang.IllegalArgumentException: Illegal Capacity: -511
	at java.util.ArrayList.<init>(ArrayList.java:157) ~[?:1.8.0_251]
	at com.hazelcast.internal.serialization.impl.SerializationUtil.readList(SerializationUtil.java:380) ~[classes/:?]
	at com.hazelcast.internal.serialization.impl.SerializationUtil.readNullableList(SerializationUtil.java:348) ~[classes/:?]
	at com.hazelcast.internal.serialization.impl.SerializationUtil.readNullableCollection(SerializationUtil.java:330) ~[classes/:?]
	at com.hazelcast.internal.partition.PartitionRuntimeState.readData(PartitionRuntimeState.java:230) ~[classes/:?]
	at com.hazelcast.internal.serialization.impl.DataSerializableSerializer.readInternal(DataSerializableSerializer.java:160) ~[classes/:?]
	at com.hazelcast.internal.serialization.impl.DataSerializableSerializer.read(DataSerializableSerializer.java:106) ~[classes/:?]
	at com.hazelcast.internal.serialization.impl.DataSerializableSerializer.read(DataSerializableSerializer.java:51) ~[classes/:?]
	at com.hazelcast.internal.serialization.impl.StreamSerializerAdapter.read(StreamSerializerAdapter.java:44) ~[classes/:?]
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.readObject(AbstractSerializationService.java:286) ~[classes/:?]
	at com.hazelcast.internal.serialization.impl.ByteArrayObjectDataInput.readObject(ByteArrayObjectDataInput.java:567) ~[classes/:?]
	at com.hazelcast.internal.cluster.impl.operations.MembersUpdateOp.readInternalImpl(MembersUpdateOp.java:117) ~[classes/:?]
	at com.hazelcast.internal.cluster.impl.operations.FinalizeJoinOp.readInternalImpl(FinalizeJoinOp.java:169) ~[classes/:?]
	at com.hazelcast.internal.cluster.impl.operations.MembersUpdateOp.readInternal(MembersUpdateOp.java:123) ~[classes/:?]
	at com.hazelcast.spi.impl.operationservice.Operation.readData(Operation.java:750) ~[classes/:?]
	at com.hazelcast.internal.serialization.impl.DataSerializableSerializer.readInternal(DataSerializableSerializer.java:160) ~[classes/:?]
	at com.hazelcast.internal.serialization.impl.DataSerializableSerializer.read(DataSerializableSerializer.java:106) ~[classes/:?]
	at com.hazelcast.internal.serialization.impl.DataSerializableSerializer.read(DataSerializableSerializer.java:51) ~[classes/:?]
	at com.hazelcast.internal.serialization.impl.StreamSerializerAdapter.read(StreamSerializerAdapter.java:44) ~[classes/:?]
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toObject(AbstractSerializationService.java:205) ~[classes/:?]
	at com.hazelcast.spi.impl.NodeEngineImpl.toObject(NodeEngineImpl.java:346) ~[classes/:?]
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:437) ~[classes/:?]
	... 4 more
```

--- a/hazelcast/src/main/java/com/hazelcast/internal/partition/MigrationInfo.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/partition/MigrationInfo.java
@@ -195,6 +195,10 @@ public class MigrationInfo implements IdentifiedDataSerializable {
         throw new IllegalStateException("Initial partition version is not set!");
     }
 
+    public UUID getUid() {
+        return uuid;
+    }
+
     @Override
     public void writeData(ObjectDataOutput out) throws IOException {
         UUIDSerializationUtil.writeUUID(out, uuid);

--- a/hazelcast/src/main/java/com/hazelcast/internal/partition/PartitionMigrationEvent.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/partition/PartitionMigrationEvent.java
@@ -16,6 +16,8 @@
 
 package com.hazelcast.internal.partition;
 
+import java.util.UUID;
+
 /**
  * An {@link java.util.EventObject} for a partition migration. Can be used by SPI services to get a callback
  * to listen to partition migration.  See {@link MigrationAwareService} for more info.
@@ -30,12 +32,15 @@ public class PartitionMigrationEvent {
 
     private final int newReplicaIndex;
 
-    public PartitionMigrationEvent(MigrationEndpoint migrationEndpoint, int partitionId,
-            int currentReplicaIndex, int newReplicaIndex) {
+    private final UUID migrationUid;
+
+    public PartitionMigrationEvent(MigrationEndpoint migrationEndpoint, int partitionId, int currentReplicaIndex,
+            int newReplicaIndex, UUID migrationUid) {
         this.migrationEndpoint = migrationEndpoint;
         this.partitionId = partitionId;
         this.currentReplicaIndex = currentReplicaIndex;
         this.newReplicaIndex = newReplicaIndex;
+        this.migrationUid = migrationUid;
     }
 
     /**
@@ -74,6 +79,13 @@ public class PartitionMigrationEvent {
      */
     public int getNewReplicaIndex() {
         return newReplicaIndex;
+    }
+
+    /**
+     * Returns the uid of the specific migration.
+     */
+    public UUID getMigrationUid() {
+        return migrationUid;
     }
 
     @Override
@@ -116,6 +128,7 @@ public class PartitionMigrationEvent {
                 + ", partitionId=" + partitionId
                 + ", currentReplicaIndex=" + currentReplicaIndex
                 + ", newReplicaIndex=" + newReplicaIndex
+                + ", uuid=" + migrationUid
                 + '}';
     }
 }

--- a/hazelcast/src/main/java/com/hazelcast/internal/partition/PartitionServiceProxy.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/partition/PartitionServiceProxy.java
@@ -215,12 +215,13 @@ public class PartitionServiceProxy implements PartitionService {
         @Override
         public Member getOwner() {
             // triggers initial partition assignment
-            final Address address = partitionService.getPartitionOwner(partitionId);
-            if (address == null) {
+            InternalPartition partition = partitionService.getPartition(partitionId);
+            PartitionReplica owner = partition.getOwnerReplicaOrNull();
+            if (owner == null) {
                 return null;
             }
 
-            return nodeEngine.getClusterService().getMember(address);
+            return nodeEngine.getClusterService().getMember(owner.address(), owner.uuid());
         }
 
         @Override

--- a/hazelcast/src/main/java/com/hazelcast/internal/partition/impl/AbstractPartitionPrimaryReplicaAntiEntropyTask.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/partition/impl/AbstractPartitionPrimaryReplicaAntiEntropyTask.java
@@ -79,6 +79,12 @@ public abstract class AbstractPartitionPrimaryReplicaAntiEntropyTask
 
         PartitionReplicaManager replicaManager = partitionService.getReplicaManager();
         replicaManager.retainNamespaces(partitionId, namespaces);
+
+        ILogger logger = nodeEngine.getLogger(getClass());
+        if (logger.isFinestEnabled()) {
+            logger.finest("Retained namespaces for partitionId=" + partitionId + ". Service namespaces="
+                    + namespaces + ", retained namespaces=" + replicaManager.getNamespaces(partitionId));
+        }
         return replicaManager.getNamespaces(partitionId);
     }
 
@@ -94,10 +100,7 @@ public abstract class AbstractPartitionPrimaryReplicaAntiEntropyTask
         for (ServiceNamespace ns : namespaces) {
             long[] versions = replicaManager.getPartitionReplicaVersions(partitionId, ns);
             long currentReplicaVersion = versions[replicaIndex - 1];
-
-            if (currentReplicaVersion > 0) {
-                versionMap.put(ns, currentReplicaVersion);
-            }
+            versionMap.put(ns, currentReplicaVersion);
         }
 
         boolean hasCallback = (callback != null);
@@ -105,6 +108,12 @@ public abstract class AbstractPartitionPrimaryReplicaAntiEntropyTask
         PartitionBackupReplicaAntiEntropyOperation op = new PartitionBackupReplicaAntiEntropyOperation(versionMap, hasCallback);
         op.setPartitionId(partitionId).setReplicaIndex(replicaIndex).setServiceName(SERVICE_NAME);
         OperationService operationService = nodeEngine.getOperationService();
+
+        ILogger logger = nodeEngine.getLogger(getClass());
+        if (logger.isFinestEnabled()) {
+            logger.finest("Sending anti-entropy operation to " + target + " for partitionId=" + partitionId
+                    + ", replicaIndex=" + replicaIndex + ", namespaces=" + versionMap);
+        }
 
         if (hasCallback) {
             operationService.createInvocationBuilder(SERVICE_NAME, op, target.address())

--- a/hazelcast/src/main/java/com/hazelcast/internal/partition/impl/InternalPartitionServiceImpl.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/partition/impl/InternalPartitionServiceImpl.java
@@ -780,8 +780,9 @@ public class InternalPartitionServiceImpl implements InternalPartitionService,
 
             if (newVersion < currentVersion) {
                 if (logger.isFinestEnabled()) {
-                    logger.finest("Already applied partition state change. Local version: " + currentVersion
-                            + ", Master version: " + newVersion + " Master: " + sender);
+                    logger.finest("Already applied partition update. partitionId=" + partitionId
+                            + ", local version: " + currentVersion
+                            + ", master version: " + newVersion + ", master: " + sender);
                 }
                 continue;
             } else if (newVersion == currentVersion) {
@@ -790,7 +791,8 @@ public class InternalPartitionServiceImpl implements InternalPartitionService,
                              + ", Received: " + newPartition);
                 }
                 if (logger.isFinestEnabled()) {
-                    logger.finest("Already applied partition state change. Version: " + currentVersion + ", Master: " + sender);
+                    logger.finest("Already applied partition update. partitionId=" + partitionId
+                            + ", version: " + currentVersion + ", master: " + sender);
                 }
                 accepted = true;
                 continue;
@@ -1442,7 +1444,7 @@ public class InternalPartitionServiceImpl implements InternalPartitionService,
             activeMigration.setStatus(migration.getStatus());
             migrationManager.finalizeMigration(migration);
             if (logger.isFineEnabled()) {
-                logger.fine("Committed " + migration + " on destination with partition state version: " + finalVersion);
+                logger.fine("Committed " + migration + " on destination with partition version: " + finalVersion);
             }
             return true;
         } finally {

--- a/hazelcast/src/main/java/com/hazelcast/internal/partition/impl/MigrationManager.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/partition/impl/MigrationManager.java
@@ -239,6 +239,9 @@ public class MigrationManager {
                 op.setPartitionId(partitionId).setNodeEngine(nodeEngine).setValidateTarget(false).setService(partitionService);
                 registerFinalizingMigration(migrationInfo);
                 OperationServiceImpl operationService = nodeEngine.getOperationService();
+                if (logger.isFineEnabled()) {
+                    logger.fine("Finalizing " + migrationInfo);
+                }
                 if (operationService.isRunAllowed(op)) {
                     // When migration finalization is triggered by subsequent migrations
                     // on partition thread, finalization may run directly on calling thread.
@@ -253,6 +256,9 @@ public class MigrationManager {
             } else {
                 PartitionReplica partitionOwner = partitionStateManager.getPartitionImpl(partitionId).getOwnerReplicaOrNull();
                 if (localReplica.equals(partitionOwner)) {
+                    // This is the primary owner of the partition and the migration was a backup replica migration.
+                    // In this case, primary owner has nothing to do anymore,
+                    // just remove the active migration and clear the migrating flag.
                     removeActiveMigration(migrationInfo);
                     partitionStateManager.clearMigratingFlag(partitionId);
                 } else {
@@ -347,6 +353,9 @@ public class MigrationManager {
             MigrationInfo activeMigrationInfo = getActiveMigration(migrationInfo.getPartitionId());
             if (migrationInfo.equals(activeMigrationInfo)) {
                 activeMigrationInfo.setStatus(migrationInfo.getStatus());
+                if (logger.isFineEnabled()) {
+                    logger.fine("Scheduled finalization of " + activeMigrationInfo);
+                }
                 finalizeMigration(activeMigrationInfo);
                 return;
             }
@@ -354,7 +363,29 @@ public class MigrationManager {
             PartitionReplica source = migrationInfo.getSource();
             if (source != null && migrationInfo.getSourceCurrentReplicaIndex() > 0
                     && source.isIdentical(node.getLocalMember())) {
-                // Finalize migration on old backup replica owner
+                // This is former backup replica owner.
+                // Former backup owner does not participate in migration transaction, data always copied
+                // from the primary replica. Former backup replica is not notified about this migration
+                // until the migration is committed on destination. Active migration is not set
+                // for this migration.
+
+                // This path can be executed multiple times,
+                // when a periodic update (latest completed migrations or the whole partition table) is received
+                // and a new migration request is submitted concurrently.
+                // That's why, migration should be validated by partition table, to determine whether
+                // this migration finalization is already processed or not.
+                InternalPartitionImpl partition = partitionStateManager.getPartitionImpl(migrationInfo.getPartitionId());
+
+                if (migrationInfo.getStatus() == MigrationStatus.SUCCESS
+                        && migrationInfo.getSourceNewReplicaIndex() != partition.getReplicaIndex(source)) {
+                    if (logger.isFinestEnabled()) {
+                        logger.finest("Already finalized " + migrationInfo + " on former backup replica. -> " + partition);
+                    }
+                    return;
+                }
+                if (logger.isFineEnabled()) {
+                    logger.fine("Scheduled finalization of " + migrationInfo + " on former backup replica.");
+                }
                 finalizeMigration(migrationInfo);
             }
         } finally {

--- a/hazelcast/src/main/java/com/hazelcast/internal/partition/impl/PartitionReplicaFragmentVersions.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/partition/impl/PartitionReplicaFragmentVersions.java
@@ -100,9 +100,7 @@ final class PartitionReplicaFragmentVersions {
     }
 
     void clear() {
-        for (int i = 0; i < versions.length; i++) {
-            versions[i] = 0;
-        }
+        Arrays.fill(versions, 0);
         dirty = false;
     }
 

--- a/hazelcast/src/main/java/com/hazelcast/internal/partition/impl/PartitionReplicaManager.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/partition/impl/PartitionReplicaManager.java
@@ -491,13 +491,11 @@ public class PartitionReplicaManager implements PartitionReplicaVersionManager {
     }
 
     @Override
-    public Collection<ServiceNamespace> getNamespaces(
-            int partitionId) {
+    public Collection<ServiceNamespace> getNamespaces(int partitionId) {
         return replicaVersions[partitionId].getNamespaces();
     }
 
-    public void retainNamespaces(int partitionId, Collection<
-            ServiceNamespace> namespaces) {
+    public void retainNamespaces(int partitionId, Collection<ServiceNamespace> namespaces) {
         PartitionReplicaVersions versions = replicaVersions[partitionId];
         versions.retainNamespaces(namespaces);
     }

--- a/hazelcast/src/main/java/com/hazelcast/internal/partition/impl/PartitionReplicaVersions.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/partition/impl/PartitionReplicaVersions.java
@@ -19,6 +19,7 @@ package com.hazelcast.internal.partition.impl;
 import com.hazelcast.internal.services.ServiceNamespace;
 
 import java.util.Collection;
+import java.util.Collections;
 import java.util.HashMap;
 import java.util.Map;
 
@@ -90,7 +91,7 @@ final class PartitionReplicaVersions {
     }
 
     Collection<ServiceNamespace> getNamespaces() {
-        return fragmentVersionsMap.keySet();
+        return Collections.unmodifiableCollection(fragmentVersionsMap.keySet());
     }
 
     @Override

--- a/hazelcast/src/main/java/com/hazelcast/internal/partition/operation/AbstractPromotionOperation.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/partition/operation/AbstractPromotionOperation.java
@@ -39,7 +39,8 @@ abstract class AbstractPromotionOperation extends AbstractPartitionOperation
     }
 
     PartitionMigrationEvent getPartitionMigrationEvent() {
-        return new PartitionMigrationEvent(DESTINATION, getPartitionId(), migrationInfo.getDestinationCurrentReplicaIndex(), 0);
+        return new PartitionMigrationEvent(DESTINATION, getPartitionId(), migrationInfo.getDestinationCurrentReplicaIndex(),
+                0, migrationInfo.getUid());
     }
 
     @Override

--- a/hazelcast/src/main/java/com/hazelcast/internal/partition/operation/FinalizeMigrationOperation.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/partition/operation/FinalizeMigrationOperation.java
@@ -133,7 +133,8 @@ public final class FinalizeMigrationOperation extends AbstractPartitionOperation
                 endpoint == MigrationEndpoint.SOURCE
                         ? migrationInfo.getSourceCurrentReplicaIndex() : migrationInfo.getDestinationCurrentReplicaIndex(),
                 endpoint == MigrationEndpoint.SOURCE
-                        ? migrationInfo.getSourceNewReplicaIndex() : migrationInfo.getDestinationNewReplicaIndex());
+                        ? migrationInfo.getSourceNewReplicaIndex() : migrationInfo.getDestinationNewReplicaIndex(),
+                migrationInfo.getUid());
     }
 
     /** Updates the replica versions on the migration source if the replica index has changed. */

--- a/hazelcast/src/main/java/com/hazelcast/internal/partition/operation/MigrationOperation.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/partition/operation/MigrationOperation.java
@@ -158,15 +158,13 @@ public class MigrationOperation extends BaseMigrationOperation implements Target
                 replicaManager.setPartitionReplicaVersions(migrationInfo.getPartitionId(), namespace,
                                                            replicaVersions, replicaOffset);
                 if (logger.isFinestEnabled()) {
-                    logger.finest("ReplicaVersions are set after migration. partitionId="
-                            + migrationInfo.getPartitionId() + " namespace: " + namespace
-                            + " replicaVersions=" + Arrays.toString(replicaVersions));
+                    logger.finest("ReplicaVersions are set after migration. " + migrationInfo
+                            + ", namespace=" + namespace + ", replicaVersions=" + Arrays.toString(replicaVersions));
                 }
             }
 
         } else if (logger.isFinestEnabled()) {
-            logger.finest("ReplicaVersions are not set since migration failed. partitionId="
-                    + migrationInfo.getPartitionId());
+            logger.finest("ReplicaVersions are not set since migration failed. " + migrationInfo);
         }
     }
 
@@ -183,7 +181,7 @@ public class MigrationOperation extends BaseMigrationOperation implements Target
     protected PartitionMigrationEvent getMigrationEvent() {
         return new PartitionMigrationEvent(MigrationEndpoint.DESTINATION,
                 migrationInfo.getPartitionId(), migrationInfo.getDestinationCurrentReplicaIndex(),
-                migrationInfo.getDestinationNewReplicaIndex());
+                migrationInfo.getDestinationNewReplicaIndex(), migrationInfo.getUid());
     }
 
     @Override

--- a/hazelcast/src/main/java/com/hazelcast/internal/partition/operation/MigrationRequestOperation.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/partition/operation/MigrationRequestOperation.java
@@ -243,7 +243,7 @@ public class MigrationRequestOperation extends BaseMigrationOperation {
     protected PartitionMigrationEvent getMigrationEvent() {
         return new PartitionMigrationEvent(MigrationEndpoint.SOURCE,
                 migrationInfo.getPartitionId(), migrationInfo.getSourceCurrentReplicaIndex(),
-                migrationInfo.getSourceNewReplicaIndex());
+                migrationInfo.getSourceNewReplicaIndex(), migrationInfo.getUid());
     }
 
     @Override

--- a/hazelcast/src/main/java/com/hazelcast/internal/partition/operation/PartitionBackupReplicaAntiEntropyOperation.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/partition/operation/PartitionBackupReplicaAntiEntropyOperation.java
@@ -16,17 +16,20 @@
 
 package com.hazelcast.internal.partition.operation;
 
+import com.hazelcast.cluster.Address;
 import com.hazelcast.internal.partition.InternalPartitionService;
+import com.hazelcast.internal.partition.PartitionReplica;
 import com.hazelcast.internal.partition.ReplicaErrorLogger;
+import com.hazelcast.internal.partition.impl.InternalPartitionImpl;
 import com.hazelcast.internal.partition.impl.InternalPartitionServiceImpl;
 import com.hazelcast.internal.partition.impl.PartitionReplicaManager;
+import com.hazelcast.internal.services.ServiceNamespace;
 import com.hazelcast.logging.ILogger;
 import com.hazelcast.nio.ObjectDataInput;
 import com.hazelcast.nio.ObjectDataOutput;
-import com.hazelcast.spi.impl.operationservice.PartitionAwareOperation;
-import com.hazelcast.internal.services.ServiceNamespace;
 import com.hazelcast.spi.impl.AllowedDuringPassiveState;
 import com.hazelcast.spi.impl.NodeEngineImpl;
+import com.hazelcast.spi.impl.operationservice.PartitionAwareOperation;
 
 import java.io.IOException;
 import java.util.HashMap;
@@ -63,8 +66,34 @@ public final class PartitionBackupReplicaAntiEntropyOperation
         int partitionId = getPartitionId();
         int replicaIndex = getReplicaIndex();
 
+        InternalPartitionImpl partition = partitionService.getPartitionStateManager().getPartitionImpl(partitionId);
+        int currentReplicaIndex = partition.getReplicaIndex(PartitionReplica.from(getNodeEngine().getLocalMember()));
+
+        ILogger logger = getLogger();
+        if (replicaIndex != currentReplicaIndex) {
+            logger.fine("Anti-entropy operation for partitionId=" + getPartitionId() + ", replicaIndex=" + getReplicaIndex()
+                    + " is received, but this node is not the expected backup replica!"
+                    + " Current replicaIndex=" + currentReplicaIndex);
+            response = false;
+            return;
+        }
+
+        Address ownerAddress = partition.getOwnerOrNull();
+        if (!getCallerAddress().equals(ownerAddress)) {
+            logger.fine("Anti-entropy operation for partitionId=" + getPartitionId() + ", replicaIndex=" + getReplicaIndex()
+                    + " is received from " + getCallerAddress() + ", but it's not the known primary replica owner: "
+                    + ownerAddress);
+            response = false;
+            return;
+        }
+
         PartitionReplicaManager replicaManager = partitionService.getReplicaManager();
         replicaManager.retainNamespaces(partitionId, versions.keySet());
+
+        if (logger.isFinestEnabled()) {
+            logger.finest("Retained namespaces for partitionId=" + partitionId + ", replicaIndex=" + replicaIndex
+                    + ". Namespaces=" + replicaManager.getNamespaces(partitionId));
+        }
 
         Iterator<Map.Entry<ServiceNamespace, Long>> iter = versions.entrySet().iterator();
         while (iter.hasNext()) {

--- a/hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/SerializationUtil.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/SerializationUtil.java
@@ -194,11 +194,16 @@ public final class SerializationUtil {
     }
 
     public static <K, V> void writeMap(@Nonnull Map<K, V> map, ObjectDataOutput out) throws IOException {
-        out.writeInt(map.size());
+        int size = map.size();
+        out.writeInt(size);
+
+        int k = 0;
         for (Map.Entry<K, V> entry : map.entrySet()) {
             out.writeObject(entry.getKey());
             out.writeObject(entry.getValue());
+            k++;
         }
+        assert size == k : "Map has been updated during serialization! Initial size: " + size + ", written size: " + k;
     }
 
     /**
@@ -277,10 +282,15 @@ public final class SerializationUtil {
      * @throws IOException when an error occurs while writing to the output
      */
     public static <T> void writeCollection(Collection<T> items, ObjectDataOutput out) throws IOException {
-        out.writeInt(items.size());
+        int size = items.size();
+        out.writeInt(size);
+
+        int k = 0;
         for (T item : items) {
             out.writeObject(item);
+            k++;
         }
+        assert size == k : "Collection has been updated during serialization! Initial size: " + size + ", written size: " + k;
     }
 
     /**

--- a/hazelcast/src/test/java/com/hazelcast/internal/partition/MigrationCorrectnessTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/internal/partition/MigrationCorrectnessTest.java
@@ -46,11 +46,8 @@ public class MigrationCorrectnessTest extends AbstractMigrationCorrectnessTest {
                 {1, 2, true},
                 {1, 2, false},
                 {2, 3, true},
-                // These variants are failing frequently.
-                // Ignored these below until we find the cause and fix.
-                // See https://github.com/hazelcast/hazelcast/issues/17377
-                // {3, 4, true},
-                // {3, 4, false},
+                {3, 4, true},
+                {3, 4, false},
         });
     }
 }

--- a/hazelcast/src/test/java/com/hazelcast/internal/partition/PartitionCorrectnessTestSupport.java
+++ b/hazelcast/src/test/java/com/hazelcast/internal/partition/PartitionCorrectnessTestSupport.java
@@ -326,8 +326,8 @@ public abstract class PartitionCorrectnessTestSupport extends HazelcastTestSuppo
                     PartitionReplicaVersionsView backupReplicaVersionsView
                             = getPartitionReplicaVersionsView(backupNode, partitionId);
                     long[] backupReplicaVersions = backupReplicaVersionsView.getVersions(namespace);
-                    assertNotNull("Versions null on " + backupNode.address + ", partitionId: " + partitionId,
-                            backupReplicaVersions);
+                    assertNotNull(namespace + " replica versions are null on " + backupNode.address
+                                    + ", partitionId: " + partitionId + ", replicaIndex: " + replica, backupReplicaVersions);
 
                     for (int i = replica - 1; i < actualBackupCount; i++) {
                         assertEquals("Replica version mismatch! Owner: " + thisAddress + ", Backup: " + address

--- a/hazelcast/src/test/java/com/hazelcast/internal/partition/service/TestAbstractMigrationAwareService.java
+++ b/hazelcast/src/test/java/com/hazelcast/internal/partition/service/TestAbstractMigrationAwareService.java
@@ -42,7 +42,7 @@ public abstract class TestAbstractMigrationAwareService<N> implements ManagedSer
 
     public volatile int backupCount;
 
-    private volatile ILogger logger;
+    protected volatile ILogger logger;
 
     @Override
     public void init(NodeEngine nodeEngine, Properties properties) {

--- a/hazelcast/src/test/java/com/hazelcast/map/impl/query/QueryRunnerTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/map/impl/query/QueryRunnerTest.java
@@ -39,6 +39,7 @@ import org.junit.experimental.categories.Category;
 import org.junit.runner.RunWith;
 
 import java.util.Set;
+import java.util.UUID;
 
 import static com.hazelcast.test.Accessors.getNodeEngineImpl;
 import static com.hazelcast.test.Accessors.getSerializationService;
@@ -101,7 +102,7 @@ public class QueryRunnerTest extends HazelcastTestSupport {
         map.addIndex(IndexType.HASH, "this");
         Predicate predicate = new EqualPredicate("this", value);
 
-        mapService.beforeMigration(new PartitionMigrationEvent(MigrationEndpoint.SOURCE, partitionId, 0, 1));
+        mapService.beforeMigration(new PartitionMigrationEvent(MigrationEndpoint.SOURCE, partitionId, 0, 1, UUID.randomUUID()));
 
         Query query = Query.of().mapName(map.getName()).predicate(predicate).iterationType(IterationType.ENTRY).build();
         QueryResult result = (QueryResult) queryRunner.runIndexOrPartitionScanQueryOnOwnedPartitions(query);
@@ -116,7 +117,8 @@ public class QueryRunnerTest extends HazelcastTestSupport {
             @Override
             public Set<QueryableEntry> filter(QueryContext queryContext) {
                 // start a new migration while executing an indexed query
-                mapService.beforeMigration(new PartitionMigrationEvent(MigrationEndpoint.SOURCE, partitionId, 0, 1));
+                mapService.beforeMigration(new PartitionMigrationEvent(MigrationEndpoint.SOURCE, partitionId, 0, 1,
+                        UUID.randomUUID()));
                 return super.filter(queryContext);
             }
         };
@@ -129,7 +131,7 @@ public class QueryRunnerTest extends HazelcastTestSupport {
     public void verifyFullScanFailureWhileMigrating() {
         Predicate predicate = new EqualPredicate("this", value);
 
-        mapService.beforeMigration(new PartitionMigrationEvent(MigrationEndpoint.SOURCE, partitionId, 0, 1));
+        mapService.beforeMigration(new PartitionMigrationEvent(MigrationEndpoint.SOURCE, partitionId, 0, 1, UUID.randomUUID()));
 
         Query query = Query.of().mapName(map.getName()).predicate(predicate).iterationType(IterationType.ENTRY).build();
         QueryResult result = (QueryResult) queryRunner.runIndexOrPartitionScanQueryOnOwnedPartitions(query);
@@ -142,7 +144,8 @@ public class QueryRunnerTest extends HazelcastTestSupport {
             @Override
             protected boolean applyForSingleAttributeValue(Comparable attributeValue) {
                 // start a new migration while executing a full scan
-                mapService.beforeMigration(new PartitionMigrationEvent(MigrationEndpoint.SOURCE, partitionId, 0, 1));
+                mapService.beforeMigration(new PartitionMigrationEvent(MigrationEndpoint.SOURCE, partitionId, 0, 1,
+                        UUID.randomUUID()));
                 return super.applyForSingleAttributeValue(attributeValue);
             }
         };

--- a/hazelcast/src/test/java/com/hazelcast/ringbuffer/impl/RingbufferServiceTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/ringbuffer/impl/RingbufferServiceTest.java
@@ -28,6 +28,8 @@ import org.junit.Test;
 import org.junit.experimental.categories.Category;
 import org.junit.runner.RunWith;
 
+import java.util.UUID;
+
 import static com.hazelcast.internal.partition.MigrationEndpoint.DESTINATION;
 import static com.hazelcast.test.Accessors.getNodeEngineImpl;
 import static org.junit.Assert.assertEquals;
@@ -49,7 +51,7 @@ public class RingbufferServiceTest extends HazelcastTestSupport {
     public void rollbackMigration() {
         Ringbuffer ringbuffer = hz.getRingbuffer("foo");
         int partitionId = getPartitionId(hz, ringbuffer.getName());
-        PartitionMigrationEvent partitionEvent = new PartitionMigrationEvent(DESTINATION, partitionId, -1, 0);
+        PartitionMigrationEvent partitionEvent = new PartitionMigrationEvent(DESTINATION, partitionId, -1, 0, UUID.randomUUID());
 
         service.rollbackMigration(partitionEvent);
 

